app/routes.py

```diff
@@ -1,2 +1,217 @@
+    tasks_response = [(task.to_json()) for task in tasks]
+    return make_response(jsonify(tasks_response), 200)
+
+
+# WAVE 1
+@tasks_bp.route("/<task_id>", methods=["GET"], strict_slashes=False)
+def get_one_task(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return make_response("", 404)
+# thank you audrey!
+    elif task.goal_id is None:
+        return jsonify({"task": task.to_json()}), 200
+    else:
+        return jsonify({"task": task.with_goal()}), 200
+
+
+@tasks_bp.route("/<task_id>", methods=["PUT"], strict_slashes=False)
+def update_task(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return make_response("", 404)
+    else:
+        form_data = request.get_json()
+        task.title = form_data["title"]
+        task.description = form_data["description"]
+        task.completed_at = form_data["completed_at"]
+        db.session.commit()
+    return jsonify({"task": task.to_json()}), 200
+
+
+@tasks_bp.route("/<task_id>", methods=["DELETE"], strict_slashes=False)
+def delete_task(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return make_response("", 404)
+    else:
+        db.session.delete(task)
+        db.session.commit()
+        task_response = {
+            "details": f'Task {task.task_id} "{task.title}" successfully deleted'}
+    return make_response(task_response), 200
+
+
+# WAVE 3
+@tasks_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"], strict_slashes=False)
+def handle_incomplete(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return make_response("", 404)
+    else:
+        task.completed_at = None
+        db.session.commit()
+        return jsonify({"task": task.to_json()}), 200
+
+
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"], strict_slashes=False)
+def handle_complete(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return make_response("", 404)
+    else:
+        task.completed_at = datetime.now()
+        db.session.commit()
+        call_slack_api(task)
+    return jsonify({"task": task.to_json()}), 200
+
+
+# WAVE 4
+def call_slack_api(task):
+    SLACK_TOKEN = os.environ.get("SLACK_BOT_TOKEN")
+    url = "https://slack.com/api/chat.postMessage"
+    payload = {
+        "channel": "task-notifications",
+        "text": f"Someone just completed the task {task.title}"}
+    headers = {
+        "Authorization": f"Bearer {SLACK_TOKEN}",
+    }
+    return requests.request("POST", url, headers=headers, data=payload)
+
+# IGNORE - WORKS BUT USES SLACK BOLT/PYTHON SDK (from slack API docs) instead of requests.package
+# def call_slack_api(task):
+#     client = WebClient(token=os.environ.get("SLACK_BOT_TOKEN"))
+#     channel_id = "task-notifications"
+#     # try:
+#     result = client.chat_postMessage(
+#         channel=channel_id,
+#         text=f"Someone just completed the task {task.title}")
+#     return result
```

Comment:
Glad you found an alternate way to send a Slack message! Remember, though, if it's unused code you should get rid of it before final submission